ext/har/types.go

```diff
@@ -0,0 +1,365 @@
+}
+
+type CacheEntry struct {
+	Expires    string `json:"expires,omitempty"`
+	LastAccess string `json:"lastAccess"`
+	ETag       string `json:"eTag"`
+	HitCount   int    `json:"hitCount"`
+	Comment    string `json:"comment,omitempty"`
+}
+
+type Request struct {
+	Method      string          `json:"method"`
+	Url         string          `json:"url"`
+	HttpVersion string          `json:"httpVersion"`
+	Cookies     []Cookie        `json:"cookies"`
+	Headers     []NameValuePair `json:"headers"`
+	QueryString []NameValuePair `json:"queryString"`
+	PostData    *PostData       `json:"postData,omitempty"`
+	BodySize    int64           `json:"bodySize"`
+	HeadersSize int64           `json:"headersSize"`
+}
+
+func ParseRequest(req *http.Request, captureContent bool) *Request {
+	if req == nil {
+		return nil
+	}
+	harRequest := Request{
+		Method:      req.Method,
+		Url:         req.URL.String(),
+		HttpVersion: req.Proto,
+		Cookies:     parseCookies(req.Cookies()),
+		Headers:     parseStringArrMap(req.Header),
+		QueryString: parseStringArrMap((req.URL.Query())),
+		BodySize:    req.ContentLength,
+		HeadersSize: calcHeaderSize(req.Header),
+	}
+
+	if captureContent && (req.Method == "POST" || req.Method == "PUT") {
+		harRequest.PostData = parsePostData(req)
+	}
+
+	return &harRequest
+}
+
+func (harEntry *Entry) FillIPAddress(req *http.Request) {
+	host, _, err := net.SplitHostPort(req.URL.Host)
+	if err != nil {
+		host = req.URL.Host
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		harEntry.ServerIpAddress = string(ip)
+	}
+
+	if ipaddr, err := net.LookupIP(host); err == nil {
+		for _, ip := range ipaddr {
+			if ip.To4() != nil {
+				harEntry.ServerIpAddress = ip.String()
+				return
+			}
+		}
+	}
+}
+
+func calcHeaderSize(header http.Header) int64 {
+	headerSize := 0
+	for headerName, headerValues := range header {
+		headerSize += len(headerName) + 2
+		for _, v := range headerValues {
+			headerSize += len(v)
+		}
+	}
+	return int64(headerSize)
+}
+
+func parsePostData(req *http.Request) *PostData {
+	defer func() {
+		if e := recover(); e != nil {
+			log.Printf("Error parsing request to %v: %v\n", req.URL, e)
+		}
+	}()
+
+	harPostData := new(PostData)
+	contentType := req.Header["Content-Type"]
+	if contentType == nil {
+		panic("Missing content type in request")
+	}
+	harPostData.MimeType = contentType[0]
+
+	if len(req.PostForm) > 0 {
+		for k, vals := range req.PostForm {
+			for _, v := range vals {
+				param := PostDataParam{
+					Name:  k,
+					Value: v,
+				}
+				harPostData.Params = append(harPostData.Params, param)
+			}
+		}
+	} else {
+		str, _ := ioutil.ReadAll(req.Body)
+		harPostData.Text = string(str)
+	}
+	return harPostData
+}
+
+func parseStringArrMap(stringArrMap map[string][]string) []NameValuePair {
+	index := 0
+	harQueryString := make([]NameValuePair, len(stringArrMap))
+	for k, v := range stringArrMap {
+		escapedKey, _ := url.QueryUnescape(k)
+		escapedValues, _ := url.QueryUnescape(strings.Join(v, ","))
```

Comment:
QueryUnescape will return an empty string if there is an error, in this case we should leave the original value if we are unable to unescape it, I think it's better than an empty string